File: tests/test_settings_menu.mjs
// tests/test_settings_menu.mjs
// TUI interaction tests for the AskUserQuestion settings menu.
//
// Drives the menu component factory the same way the production TUI
// would (via handleInput on synthetic key sequences), and asserts on
// the state exposed by the component for inspection. Doesn't depend
// on a real terminal — uses the same fake tui/theme pattern as
// tests/test_tui_render.mjs.

import { test } from "node:test";
import assert from "node:assert/strict";
import {
	buildSettingsMenuComponent,
	DEFAULT_SECTIONS,
} from "../src/settings-menu.ts";

function makeFakeTui() {
	return {
		requestRender: () => {},
		terminal: { rows: 24, cols: 80 },
	};
}

function makeFakeTheme() {
	const F = (_color, text) => text; // strip theme markup for stable assertions
	return {
		fg: F,
		bg: F,
		bold: (s) => s,
		italic: (s) => s,
		strikethrough: (s) => s,
	};
}

const fakeTheme = makeFakeTheme();

/**
 * Build a menu with synthetic defaults and a change recorder.
 * Returns { component, getCurrent, changes, getDone, state }.
 */
function makeMenu(overrides = {}) {
	const sections = overrides.sections ?? DEFAULT_SECTIONS;
	const current = overrides.current ?? {
		browserEnabled: true,
		browserAutoOpen: false,
		browserMinQuestions: 2,
		copyUrlToClipboard: true,
		bellOnQuestion: true,
		notificationOnQuestion: false,
		notificationDelaySeconds: 30,
		ttsOnQuestion: false,
		onQuestionCommand: "",
		heartbeatWhileActive: false,
		heartbeatIntervalMinutes: 4.5,
		debounceMs: 300,
		dangerCheckEnabled: true,
	};
	const getCurrent = overrides.getCurrent ?? (() => current);
	const changes = [];
	const tui = makeFakeTui();
	let doneValue = null;
	const factory = buildSettingsMenuComponent({
		sections,
		getCurrent,
		onChange: (id, value) => changes.push({ id, value }),
		onExit: overrides.onExit,
	});
	const component = factory(tui, fakeTheme, {}, (v) => {
		doneValue = v;
	});
	return {
		component,
		getCurrent,
		changes,
		getDone: () => doneValue,
		_state: component._state,
		getError: () => component.getError(),
	};
}

const DOWN = "\u001b[B";
const UP = "\u001b[A";
const RIGHT = "\u001b[C";
const LEFT = "\u001b[D";
const ENTER = "\r";
const ESC = "\u001b";
const SPACE = " ";

// ---- Factory shape -------------------------------------------------------

test("buildSettingsMenuComponent returns render/handleInput/dispose", () => {
	const { component } = makeMenu();
	assert.equal(typeof component.render, "function");
	assert.equal(typeof component.handleInput, "function");
	assert.equal(typeof component.dispose, "function");
	assert.equal(typeof component.invalidate, "function");
});

test("DEFAULT_SECTIONS has 5 sections covering all 13 fields", () => {
	assert.equal(DEFAULT_SECTIONS.length, 5);
	const totalSettings = DEFAULT_SECTIONS.reduce((n, s) => n + s.settings.length, 0);
	assert.equal(totalSettings, 13, "expected 13 settings across 5 sections");
	const titles = DEFAULT_SECTIONS.map((s) => s.title);
	assert.deepEqual(titles, [
		"Browser",
		"Notifications",
		"TTS & Command",
		"Heartbeat & Debounce",
		"Question display",
	]);
});

// ---- Section navigation --------------------------------------------------

test("right-arrow on section header enters the section", () => {
	const { component, _state } = makeMenu();
	assert.equal(_state().mode, "sections");
	component.handleInput(RIGHT);
	assert.equal(_state().mode, "settings");
	assert.equal(_state().sectionIndex, 0);
});

test("Enter on section header enters the section (alias for right-arrow)", () => {
	const { component, _state } = makeMenu();
	component.handleInput(ENTER);
	assert.equal(_state().mode, "settings");
});

test("left-arrow from a section returns to the section list", () => {
	const { component, _state } = makeMenu();
	component.handleInput(RIGHT); // enter first section
	assert.equal(_state().mode, "settings");
	component.handleInput(LEFT);
	assert.equal(_state().mode, "sections");
});

test("Esc on a section returns to the section list", () => {
	const { component, _state, getDone } = makeMenu();
	component.handleInput(RIGHT);
	assert.equal(_state().mode, "settings");
	component.handleInput(ESC);
	assert.equal(_state().mode, "sections");
	assert.equal(getDone(), null, "Esc inside a section must not exit the menu");
});

test("Esc at the section list exits the menu (calls done)", () => {
	const { component, getDone } = makeMenu();
	component.handleInput(ESC);
	const v = getDone();
	assert.ok(v !== null, "done() should be called on exit");
	assert.equal(v.lifecycle, "exited");
});

test("left-arrow at the section list exits the menu", () => {
	const { component, getDone } = makeMenu();
	component.handleInput(LEFT);
	const v = getDone();
	assert.ok(v !== null);
	assert.equal(v.lifecycle, "exited");
});

test("down-arrow on sections moves to the next section", () => {
	const { component, _state } = makeMenu();
	assert.equal(_state().sectionIndex, 0);
	component.handleInput(DOWN);
	assert.equal(_state().sectionIndex, 1);
	component.handleInput(DOWN);
	assert.equal(_state().sectionIndex, 2);
});

test("up-arrow wraps from index 0 to the last section", () => {
	const { component, _state } = makeMenu();
	component.handleInput(UP);
	assert.equal(_state().sectionIndex, DEFAULT_SECTIONS.length - 1);
});

test("down-arrow wraps from the last section to 0", () => {
	const { component, _state } = makeMenu();
	for (let i = 0; i < DEFAULT_SECTIONS.length; i++) component.handleInput(DOWN);
	assert.equal(_state().sectionIndex, 0);
});

// ---- Setting list navigation ---------------------------------------------

test("down-arrow in a setting list moves to the next setting", () => {
	const { component, _state } = makeMenu();
	component.handleInput(RIGHT); // enter first section (Browser)
	assert.equal(_state().settingIndex, 0);
	component.handleInput(DOWN);
	assert.equal(_state().settingIndex, 1);
});

test("up-arrow in a setting list moves to the previous setting (wraps)", () => {
	const { component, _state } = makeMenu();
	component.handleInput(RIGHT);
	assert.equal(_state().settingIndex, 0);
	component.handleInput(UP);
	assert.equal(_state().settingIndex, 3, "Browser has 4 settings; up wraps to last");
});

// ---- Boolean toggling ----------------------------------------------------

test("Space on a boolean setting toggles its value", () => {
	const { component, _state, changes } = makeMenu();
	component.handleInput(RIGHT); // enter Browser section
	component.handleInput(DOWN); // skip browserEnabled; move to browserAutoOpen
	// We're at index 1 (browserAutoOpen = false). Toggle.
	component.handleInput(SPACE);
	assert.equal(changes.length, 1);
	assert.equal(changes[0].id, "browserAutoOpen");
	assert.equal(changes[0].value, true);
	void _state;
});

test("Space on the first boolean toggles browserEnabled (true → false)", () => {
	const { component, changes } = makeMenu();
	component.handleInput(RIGHT);
	component.handleInput(SPACE);
	assert.equal(changes.length, 1);
	assert.equal(changes[0].id, "browserEnabled");
	assert.equal(changes[0].value, false);
});

test("Space on a number setting is a no-op (does not call onChange)", () => {
	const { component, changes } = makeMenu();
	// Navigate to Notifications section (index 1)
	component.handleInput(DOWN);
	component.handleInput(RIGHT);
	// Index 0 is notificationOnQuestion (boolean) — skip to index 1 (number)
	component.handleInput(DOWN);
	component.handleInput(SPACE);
	assert.equal(changes.length, 0, "Space on number must not toggle");
});

// ---- Number editor -------------------------------------------------------

test("Enter on a number setting opens an editor; Enter commits the typed value", () => {
	const { component, changes, _state } = makeMenu();
	// Navigate to Notifications section
	component.handleInput(DOWN);
	component.handleInput(RIGHT);
	// Index 1 is notificationDelaySeconds (number, default 30)
	component.handleInput(DOWN);
	assert.equal(_state().inEditor, false);
	component.handleInput(ENTER);
	assert.equal(_state().inEditor, true, "Enter should open the editor");
	// Type "60" character by character
	component.handleInput("6");
	component.handleInput("0");
	component.handleInput(ENTER);
	assert.equal(changes.length, 1);
	assert.equal(changes[0].id, "notificationDelaySeconds");
	assert.equal(changes[0].value, 60);
	assert.equal(_state().inEditor, false, "editor should close after a valid commit");
});

test("Editor rejects non-numeric input and does NOT call onChange", () => {
	const { component, changes, _state, getError } = makeMenu();
	component.handleInput(DOWN);
	component.handleInput(RIGHT); // enter Notifications
	component.handleInput(DOWN); // index 1: notificationDelaySeconds
	component.handleInput(ENTER); // open editor
	// Type "abc" — Editor accepts arbitrary chars, but the menu must
	// reject on commit.
	component.handleInput("a");
	component.handleInput("b");
	component.handleInput("c");
	component.handleInput(ENTER);
	assert.equal(changes.length, 0, "non-numeric must not commit");
	assert.equal(_state().inEditor, true, "editor stays open on invalid commit");
	const err = getError();
	assert.ok(err, "expected an error message after invalid commit");
	assert.match(err, /number|invalid/i);
});

test("Editor rejects out-of-range number (negative)", () => {
	const { component, changes, _state, getError } = makeMenu();
	component.handleInput(DOWN);
	component.handleInput(RIGHT);
	component.handleInput(DOWN); // notificationDelaySeconds
	component.handleInput(ENTER);
	component.handleInput("-");
	component.handleInput("5");
	component.handleInput(ENTER);
	assert.equal(changes.length, 0, "-5 must not commit");
	assert.equal(_state().inEditor, true);
	assert.ok(getError(), "expected an error message");
});

test("Esc in the editor cancels and does NOT call onChange", () => {
	const { component, changes, _state } = makeMenu();
	component.handleInput(DOWN);
	component.handleInput(RIGHT);
	component.handleInput(DOWN); // notificationDelaySeconds
	component.handleInput(ENTER); // open editor
	component.handleInput("9");
	component.handleInput("9");
	component.handleInput(ESC); // cancel
	assert.equal(changes.length, 0);
	assert.equal(_state().inEditor, false);
});

// ---- String editor -------------------------------------------------------

test("Enter on a string setting opens an editor; Enter commits the typed text", () => {
	const { component, changes, _state } = makeMenu();
	// Navigate to TTS & Command (section index 2)
	component.handleInput(DOWN);
	component.handleInput(DOWN);
	component.handleInput(RIGHT);
	// Index 1 is onQuestionCommand (string, default "")
	component.handleInput(DOWN);
	component.handleInput(ENTER);
	assert.equal(_state().inEditor, true);
	component.handleInput("e");
	component.handleInput("c");
	component.handleInput("h");
	component.handleInput("o");
	component.handleInput(" ");
	component.handleInput("a");
	component.handleInput("s");
	component.handleInput("k");
	component.handleInput(ENTER);
	assert.equal(changes.length, 1);
	assert.equal(changes[0].id, "onQuestionCommand");
	assert.equal(changes[0].value, "echo ask");
});

// ---- Render --------------------------------------------------------------

test("render at section level shows all 5 section titles", () => {
	const { component } = makeMenu();
	const joined = component.render(80).join("\n");
	for (const s of DEFAULT_SECTIONS) {
		assert.match(joined, new RegExp(s.title), `expected to see section: ${s.title}`);
	}
});

test("render inside a section shows each setting label and current value", () => {
	const { component } = makeMenu();
	component.handleInput(RIGHT); // enter Browser
	const joined = component.render(80).join("\n");
	// Boolean true → "on"; false → "off"
	assert.match(joined, /Browser enabled/);
	assert.match(joined, /Auto-open browser/);
	assert.match(joined, /Min questions/);
	assert.match(joined, /Copy URL to clipboard/);
	// browserMinQuestions = 2 → "2"
	assert.match(joined, /\b2\b/);
});

test("render reflects the latest getCurrent() — onChange is picked up on next render", () => {
	// getCurrent is a closure that returns whatever the latest "saved"
	// settings are. Simulating the persistence layer.
	let live = { bellOnQuestion: true };
	const { component } = makeMenu({ getCurrent: () => live });
	let joined = component.render(80).join("\n");
	assert.match(joined, /AskUserQuestion settings/);
	// Enter Bell section (Question display, index 4)
	for (let i = 0; i < 4; i++) component.handleInput(DOWN);
	component.handleInput(RIGHT);
	// Index 0 is bellOnQuestion
	const before = component.render(80).join("\n");
	assert.match(before, /on/i);
	// Toggle — the production onChange callback would call saveSettings,
	// which updates the file, which the next getSettings() reads. We
	// simulate that here by mutating live.
	component.handleInput(SPACE);
	live = { bellOnQuestion: false };
	const after = component.render(80).join("\n");
	assert.match(after, /off/i);
});

// ---- dispose -------------------------------------------------------------

test("dispose() is safe to call multiple times", () => {
	const { component } = makeMenu();
	component.dispose();
	component.dispose(); // no throw
});

// ---- onExit fires only on full exit --------------------------------------

test("onExit is called only when the user backs out at the top level", () => {
	const exits = [];
	const { component } = makeMenu({ onExit: () => exits.push(true) });
	component.handleInput(RIGHT); // enter section — no exit
	component.handleInput(LEFT); // back to sections — no exit
	component.handleInput(ESC); // exit menu
	assert.deepEqual(exits, [true]);
});

File: tests/test_settings_menu.py
"""Settings menu — data layer tests.

Exercises the persistence flow the settings menu will drive: read the
current merged view, apply a user patch, write the result via
`saveSettings` (the only persistence path), and read the file back from
disk to verify the round-trip.

The actual TUI is harder to drive from a Python test (it needs a real
terminal and key events); this suite targets the data path that the TUI
ultimately calls. The TUI itself is covered by tests/test_settings_menu.mjs.
"""
from __future__ import annotations

import json
import os
import shutil
import tempfile
from pathlib import Path

import pytest
from conftest import run_harness


# ---------------------------------------------------------------------------
# helpers
# ---------------------------------------------------------------------------


def _apply(patch: dict, cwd: str | None = None, global_dir: str | None = None) -> dict:
    """Run the harness applySettingsChange command and return the result.

    `global_dir` redirects PI_AGENT_DIR so the harness's getSettings()
    call reads from an empty temp dir (instead of polluting or being
    polluted by the real ~/.pi/agent/).
    """
    cmd: dict = {"cmd": "applySettingsChange", "patch": patch}
    if cwd is not None:
        cmd["cwd"] = cwd
    if global_dir is not None:
        cmd["globalDir"] = global_dir
    return run_harness(cmd)


@pytest.fixture
def fresh_cwd(tmp_path: Path) -> str:
    """An empty tmp dir used as cwd for the menu's save path.

    The menu writes to <cwd>/.pi/ask-user-question.json — this fixture
    gives that file a clean home per test.
    """
    return str(tmp_path)


@pytest.fixture
def empty_global_dir(tmp_path: Path) -> str:
    """A fresh tmp dir used as PI_AGENT_DIR for the harness subprocess."""
    return str(tmp_path)


# ---------------------------------------------------------------------------
# 1. boolean field round-trip
# ---------------------------------------------------------------------------


def test_apply_boolean_field_writes_file_with_that_field_set(
    fresh_cwd: str, empty_global_dir: str
):
    """A boolean patch writes the project file with that boolean set."""
    r = _apply({"bellOnQuestion": False}, cwd=fresh_cwd, global_dir=empty_global_dir)
    assert r["ok"] is True
    assert r["fileExists"] is True
    assert r["fileContent"]["bellOnQuestion"] is False


# ---------------------------------------------------------------------------
# 2. round-trip across 3 fields
# ---------------------------------------------------------------------------


def test_apply_three_changes_round_trip(fresh_cwd: str, empty_global_dir: str):
    """Apply 3 changes across 3 different fields; all 3 are correct on read-back."""
    r = _apply(
        {
            "bellOnQuestion": False,
            "notificationDelaySeconds": 60,
            "ttsOnQuestion": True,
        },
        cwd=fresh_cwd,
        global_dir=empty_global_dir,
    )
    assert r["ok"] is True
    fc = r["fileContent"]
    assert fc["bellOnQuestion"] is False
    assert fc["notificationDelaySeconds"] == 60
    assert fc["ttsOnQuestion"] is True
    # Other defaults are preserved in the file (full-merged-view write).
    assert fc["browserEnabled"] is True
    assert fc["dangerCheckEnabled"] is True
    # After-view (getSettings) should reflect all three.
    assert r["after"]["bellOnQuestion"] is False
    assert r["after"]["notificationDelaySeconds"] == 60
    assert r["after"]["ttsOnQuestion"] is True


# ---------------------------------------------------------------------------
# 3. global-only field override creates project file
# ---------------------------------------------------------------------------


def test_apply_change_to_global_only_field_creates_project_file(
    fresh_cwd: str, empty_global_dir: str
):
    """A patch that overrides a default field creates <cwd>/.pi/ask-user-question.json."""
    # bellOnQuestion defaults to true. Setting it to false must write the file.
    project_path = os.path.join(fresh_cwd, ".pi", "ask-user-question.json")
    assert not os.path.exists(project_path), "fixture left a stale file behind"

    r = _apply({"bellOnQuestion": False}, cwd=fresh_cwd, global_dir=empty_global_dir)
    assert r["ok"] is True
    assert os.path.exists(project_path), "project file was not created"
    assert r["projectPath"] == project_path


# ---------------------------------------------------------------------------
# 4. partial patch preserves other fields
# ---------------------------------------------------------------------------


def test_apply_partial_patch_preserves_other_fields_in_file(
    fresh_cwd: str, empty_global_dir: str
):
    """A 1-field patch writes a file that contains ALL 13 fields (full merged view).

    The patch is conceptually `{ bellOnQuestion: false }`, but the file
    is written with the full merged view: every field is present, with
    the patched value where overridden and the merged value elsewhere.
    """
    r = _apply({"bellOnQuestion": False}, cwd=fresh_cwd, global_dir=empty_global_dir)
    assert r["ok"] is True
    fc = r["fileContent"]
    # All 13 fields must be present.
    expected_fields = {
        "browserEnabled",
        "browserAutoOpen",
        "browserMinQuestions",
        "copyUrlToClipboard",
        "bellOnQuestion",
        "notificationOnQuestion",
        "notificationDelaySeconds",
        "ttsOnQuestion",
        "onQuestionCommand",
        "heartbeatWhileActive",
        "heartbeatIntervalMinutes",
        "debounceMs",
        "dangerCheckEnabled",
    }
    assert set(fc.keys()) == expected_fields, (
        f"file keys {set(fc.keys())} != expected {expected_fields}"
    )
    # The patched field is overridden.
    assert fc["bellOnQuestion"] is False
    # Other fields keep their default values (since no global + no prior
    # project file existed before this test).
    assert fc["browserEnabled"] is True
    assert fc["browserMinQuestions"] == 2
    assert fc["notificationDelaySeconds"] == 30


def test_apply_partial_patch_preserves_pre_existing_field_values(
    fresh_cwd: str, empty_global_dir: str
):
    """A patch applied on top of an existing project file does not blow away
    fields that were already set there."""
    project_dir = os.path.join(fresh_cwd, ".pi")
    os.makedirs(project_dir, exist_ok=True)
    project_path = os.path.join(project_dir, "ask-user-question.json")
    # Pre-existing file: browserEnabled = false (an explicit override).
    with open(project_path, "w") as f:
        json.dump({"browserEnabled": False}, f)

    r = _apply({"bellOnQuestion": False}, cwd=fresh_cwd, global_dir=empty_global_dir)
    assert r["ok"] is True
    fc = r["fileContent"]
    # Pre-existing override is preserved.
    assert fc["browserEnabled"] is False
    # New patch is applied.
    assert fc["bellOnQuestion"] is False
    # Defaults are still there.
    assert fc["browserMinQuestions"] == 2


# ---------------------------------------------------------------------------
# 5. number field type preservation
# ---------------------------------------------------------------------------


def test_apply_number_writes_as_number_not_string(
    fresh_cwd: str, empty_global_dir: str
):
    """A number patch must be persisted as a JSON number, not a string."""
    r = _apply(
        {"notificationDelaySeconds": 45},
        cwd=fresh_cwd,
        global_dir=empty_global_dir,
    )
    assert r["ok"] is True
    fc = r["fileContent"]
    assert fc["notificationDelaySeconds"] == 45
    # Read raw to confirm it's not a quoted string.
    with open(r["projectPath"]) as f:
        raw = f.read()
    assert '"notificationDelaySeconds": 45' in raw or '"notificationDelaySeconds":45' in raw, (
        f"expected number literal in raw file; got: {raw[:200]}"
    )


def test_apply_fractional_number_writes_as_number(fresh_cwd: str, empty_global_dir: str):
    """heartbeatIntervalMinutes accepts fractional values (0.5–60)."""
    r = _apply(
        {"heartbeatIntervalMinutes": 2.5},
        cwd=fresh_cwd,
        global_dir=empty_global_dir,
    )
    assert r["ok"] is True
    assert r["fileContent"]["heartbeatIntervalMinutes"] == 2.5


# ---------------------------------------------------------------------------
# 6. invalid input rejection (out-of-range)
# ---------------------------------------------------------------------------


def test_apply_out_of_range_number_is_clamped_or_rejected(
    fresh_cwd: str, empty_global_dir: str
):
    """A patch that violates range constraints is rejected by the sanitizer.

    The merge layer in `getSettings()` re-validates:
    notificationDelaySeconds must be in [0, 300]. A patch of -5 is
    rejected by `sanitize()`, so the merged view falls back to the
    default (30). The contract:

    - `after` (the post-`getSettings()` view) MUST NOT contain -5.
    - The TUI is the gatekeeper that prevents -5 from being saved in
      the first place; the data layer's sanitizer is the safety net.

    This test pins down the safety-net contract.
    """
    r = _apply(
        {"notificationDelaySeconds": -5},
        cwd=fresh_cwd,
        global_dir=empty_global_dir,
    )
    # Post-sanitize view: -5 is dropped, defaults to 30.
    assert r["after"]["notificationDelaySeconds"] == 30
    # saveSettings itself is permissive (it writes the raw merged view
    # so a future bug fix in sanitize() can be observed), but the resolved
    # after-view is what consumers see via getSettings().
    # The TUI test (tests/test_settings_menu.mjs) verifies that the menu
    # refuses to dispatch -5 to onChange in the first place.


# ---------------------------------------------------------------------------
# 7. file path resolution
# ---------------------------------------------------------------------------


def test_apply_writes_to_cwd_pi_subdir_not_elsewhere(
    fresh_cwd: str, empty_global_dir: str
):
    """The file must be created at <cwd>/.pi/ask-user-question.json."""
    r = _apply({"bellOnQuestion": False}, cwd=fresh_cwd, global_dir=empty_global_dir)
    expected = os.path.join(fresh_cwd, ".pi", "ask-user-question.json")
    assert r["projectPath"] == expected
    assert os.path.exists(expected)
    # And nowhere else relevant.
    assert not os.path.exists(os.path.join(fresh_cwd, "ask-user-question.json"))
    # Global file is not touched (the menu only writes the project layer).
    global_path = os.path.join(empty_global_dir, "ask-user-question.json")
    assert not os.path.exists(global_path)


def test_apply_creates_pi_subdir_if_missing(fresh_cwd: str, empty_global_dir: str):
    """saveSettings must mkdir -p the .pi/ subdir if it doesn't exist."""
    pi_dir = os.path.join(fresh_cwd, ".pi")
    assert not os.path.exists(pi_dir)
    r = _apply({"bellOnQuestion": False}, cwd=fresh_cwd, global_dir=empty_global_dir)
    assert r["ok"] is True
    assert os.path.isdir(pi_dir)
    assert os.path.isfile(os.path.join(pi_dir, "ask-user-question.json"))


# ---------------------------------------------------------------------------
# 8. empty patch is a no-op write (still writes the full merged view)
# ---------------------------------------------------------------------------


def test_apply_empty_patch_writes_defaults(fresh_cwd: str, empty_global_dir: str):
    """An empty patch is allowed: writes the full default view."""
    r = _apply({}, cwd=fresh_cwd, global_dir=empty_global_dir)
    assert r["ok"] is True
    fc = r["fileContent"]
    assert fc["bellOnQuestion"] is True  # default
    assert fc["browserMinQuestions"] == 2  # default


# ---------------------------------------------------------------------------
# 9. multiple sequential patches accumulate
# ---------------------------------------------------------------------------


def test_sequential_patches_accumulate(fresh_cwd: str, empty_global_dir: str):
    """Two patches applied in sequence both persist (the second sees the first)."""
    _apply({"bellOnQuestion": False}, cwd=fresh_cwd, global_dir=empty_global_dir)
    r2 = _apply(
        {"notificationDelaySeconds": 120},
        cwd=fresh_cwd,
        global_dir=empty_global_dir,
    )
    fc = r2["fileContent"]
    assert fc["bellOnQuestion"] is False  # from first patch
    assert fc["notificationDelaySeconds"] == 120  # from second patch